### PR TITLE
Eslint Plugin: Add TypeScript as peer dependency and make it optional

### LINKED ISSUE #29940: Error: Failed to load plugin '@typescript-eslint' declared in '--config » plugin:@wordpress/recommended': Cannot find module 'typescript'
## Description

Using `@wordpress/scripts@14.0.0` and `wp-scripts lint-js`  is currently failing to lint JavaScript files because `typescript` is missing as a dependency.

## Step-by-step reproduction instructions

1. Create `package.json`
```json
{
  "name": "ts",
  "version": "0.0.0",
  "description": "",
  "main": "index.js",
  "author": "",
  "license": "private",
  "devDependencies": {
    "@wordpress/scripts": "^14.0.0"
  }
}
```
2. Create `index.js`
3. Run `npx wp-scripts lint-js index.js`

## Expected behaviour
File is linted without errors.

## Actual behaviour
```
npx wp-scripts lint-js index.js

Oops! Something went wrong! :(

ESLint: 7.22.0

Error: Failed to load plugin '@typescript-eslint' declared in '--config » plugin:@wordpress/recommended': Cannot find module 'typescript'
Require stack:
- /private/tmp/ts/node_modules/tsutils/typeguard/2.8/node.js
- /private/tmp/ts/node_modules/tsutils/typeguard/2.9/node.js
- /private/tmp/ts/node_modules/tsutils/typeguard/3.0/node.js
- /private/tmp/ts/node_modules/tsutils/typeguard/3.2/node.js
- /private/tmp/ts/node_modules/tsutils/typeguard/node.js
- /private/tmp/ts/node_modules/tsutils/typeguard/index.js
- /private/tmp/ts/node_modules/tsutils/index.js
- /private/tmp/ts/node_modules/@typescript-eslint/eslint-plugin/dist/util/isTypeReadonly.js
- /private/tmp/ts/node_modules/@typescript-eslint/eslint-plugin/dist/util/index.js
- /private/tmp/ts/node_modules/@typescript-eslint/eslint-plugin/dist/rules/adjacent-overload-signatures.js
- /private/tmp/ts/node_modules/@typescript-eslint/eslint-plugin/dist/rules/index.js
- /private/tmp/ts/node_modules/@typescript-eslint/eslint-plugin/dist/index.js
- /private/tmp/ts/node_modules/@eslint/eslintrc/lib/config-array-factory.js
- /private/tmp/ts/node_modules/@eslint/eslintrc/lib/index.js
- /private/tmp/ts/node_modules/eslint/lib/cli-engine/cli-engine.js
- /private/tmp/ts/node_modules/eslint/lib/eslint/eslint.js
- /private/tmp/ts/node_modules/eslint/lib/eslint/index.js
- /private/tmp/ts/node_modules/eslint/lib/cli.js
- /private/tmp/ts/node_modules/eslint/bin/eslint.js
Referenced from: /private/tmp/ts/node_modules/@wordpress/eslint-plugin/index.js
    at Function.Module._resolveFilename (internal/modules/cjs/loader.js:880:15)
    at Function.Module._load (internal/modules/cjs/loader.js:725:27)
    at Module.require (internal/modules/cjs/loader.js:952:19)
    at require (/private/tmp/ts/node_modules/v8-compile-cache/v8-compile-cache.js:159:20)
    at Object.<anonymous> (/private/tmp/ts/node_modules/tsutils/typeguard/2.8/node.js:7:12)
    at Module._compile (/private/tmp/ts/node_modules/v8-compile-cache/v8-compile-cache.js:192:30)
    at Object.Module._extensions..js (internal/modules/cjs/loader.js:1092:10)
    at Module.load (internal/modules/cjs/loader.js:928:32)
    at Function.Module._load (internal/modules/cjs/loader.js:769:14)
    at Module.require (internal/modules/cjs/loader.js:952:19)
````

I'm using node v14.15.4 with npm 6.14.11.

Probably introduced by #27143.

### PR DESCRIPTION
**Pull request by sarayourfriend**
## Description
Adds an explicit dependency on TypeScript to `eslint-plugin`.

See https://github.com/WordPress/gutenberg/issues/29940 for details.

Fixes #29940

## How has this been tested?
1. Checkout this branch
2. Run `npm run build`
3. Follow the reproduction steps in #29940 but point `scripts` to `file:../gutenberg/packages/scripts` (assuming `gutenberg` lives in the same repository as the test folder for the test `package.json`)
For example:

```json
{
  "name": "ts",
  "version": "0.0.0",
  "description": "",
  "main": "index.js",
  "author": "",
  "license": "private",
  "devDependencies": {
    "@wordpress/scripts": "file:../gutenberg/packages/scripts"
  }
}
```

There should be no errors when running `npx wp-scripts lint-js index.js`

You must `npm install` in your test directory before running `npx ...`

## Types of changes
Bug fix

## Checklist:
- [x] My code is tested.
- [x] My code follows the WordPress code style. 
- [N/A] My code follows the accessibility standards. 
- [N/A] I've tested my changes with keyboard and screen readers. 
- [x] My code has proper inline documentation. 
- [x] I've included developer documentation if appropriate. 
- [N/A] I've updated all React Native files affected by any refactorings/renamings in this PR.

### CODE REVIEW

**Comment by gziolo** [comment]
Huh, this is what I see at https://www.npmjs.com/package/typescript

Screen Shot 2021-03-17 at 18 58 16

I guess we can't do much about it.

**Comment by ocean90** [comment]
Is it possible to create a config which treats *.js and *.ts files separately so we can make TS a peer dependency like eslint? Maybe at least until TS is more common in the project?

Basically if a project has no *.ts files don't require TS to be installed.

File: packages/eslint-plugin/package.json [hunk 1]
```diff
@@ -46,7 +46,8 @@
 		"eslint-plugin-react-hooks": "^4.2.0",
 		"globals": "^12.0.0",
 		"prettier": "npm:wp-prettier@2.2.1-beta-1",
-		"requireindex": "^1.2.0"
+		"requireindex": "^1.2.0",
+		"typescript": "^4.1.3"
```
Comment (sirreal) [inline packages/eslint-plugin/package.json]: I don't think this is a good idea. TypeScript should be a peer dependency if we need to require it. It should be provided by the consumer package and not use the version we require. We don't want duplicates.

If we require packages that have this peer dependency, we should either pass the peer dep along or reconsider whether we can remove that requirement.

**Comment by sirreal** [comment]
> Is it possible to create a config which treats *.js and *.ts files separately so we can make TS a peer dependency like eslint? Maybe at least until TS is more common in the project?
> 
> Basically if a project has no *.ts files don't require TS to be installed.

This is worth exploring, it seems like a straightforward way to address the issue.

**Comment by gziolo** [comment]
@sirreal and @ocean90 - I applied some iterations to this PR and now TypeScript should be optional, at least with npm 6. I think npm 7 installs peer dependencies by default so we can look at the proposed approach separately to address it.

**Comment by gziolo** [comment]
This is interesting way to declare peer dependencies:
https://github.com/typescript-eslint/typescript-eslint/blob/9ed3c5020cf390f7f26cdb24638c222fc0383647/packages/parser/package.json#L58-L62

```json
"peerDependenciesMeta": {
    "typescript": {
      "optional": true
    }
  },
```

**Review by gziolo — APPROVED** [review]
I would appreciate some sanity check from someone else since I added most of the code changes 😅 

I plan to cherry-pick this PR and publish to npm as soon as it gets accepted.

**Comment by ocean90** [comment]
Based on the comment https://github.com/npm/cli/issues/1247#issuecomment-688360885, I think the [`peerDependenciesMeta`](https://docs.npmjs.com/cli/v7/configuring-npm/package-json#peerdependenciesmeta) is required so that npm v7 doesn't auto-resolve optional peerDependencies?

This is different for `eslint` which is a required peer dependency.

**Comment by gziolo** [comment]
> Based on the comment [npm/cli#1247 (comment)](https://github.com/npm/cli/issues/1247#issuecomment-688360885), I think the [`peerDependenciesMeta`](https://docs.npmjs.com/cli/v7/configuring-npm/package-json#peerdependenciesmeta) is required so that npm v7 doesn't auto-resolve optional peerDependencies?
> 
> This is different for `eslint` which is a required peer dependency.

Yes, it looks like you are correct and it was implemented as noted it https://github.com/npm/rfcs/issues/221.

Detailed documentation is here:
https://docs.npmjs.com/cli/v7/configuring-npm/package-json#peerdependenciesmeta

**Comment by gziolo** [comment]
`typescript` is now marked for npm 7+ as optional with https://github.com/WordPress/gutenberg/pull/29942/commits/ef5c6ca081ea21b890336e02d7a7413d2f99884c.

**Review by ocean90 — APPROVED** [review]
Tested with and without TS and it's working as expected, should be good to merge.

**Comment by gziolo** [comment]
Thank you all for help 🙇🏻

**Pull request merged**